### PR TITLE
Background processing

### PR DESCRIPTION
**Pull request by morgoth**
Very basic implementation for background processing support using ActvieJob.

It requires more work, but I would like to start collecting your feedback about implementation and possible features.

I tested it with file and s3 storages locally.

Questions:
- ~~I think it make sense to add db column where we can mark that file is being processed () - do you think boolean `:attachment_name:_processing_in_background` is fine?~~ done
- To force processing in background one have to set array of styles in `process_in_background` options, but also set `only_processing` to fake `[:none]`, as empty array processes all by default. I would like to solve it some better way. How it should behave?

[ref #2260]

### CODE REVIEW

File: lib/paperclip/attachment.rb [hunk 1]
```diff
@@ -383,6 +386,21 @@ def instance_read(attr)
       end
     end
 
+    def generate_style_files(*styles)
+      @file = Paperclip.io_adapters.for(self)
+      @queued_for_write[:original] = @file
```
Comment (morgoth) [inline lib/paperclip/attachment.rb]: It seems to be little hacky, but without this other styles won't process

File: lib/paperclip/has_attached_file.rb [hunk 2]
```diff
@@ -100,6 +101,20 @@ def add_active_record_callbacks
       end
     end
 
+    def add_background_processing
+      require "paperclip/process_job"
```
Comment (morgoth) [inline lib/paperclip/has_attached_file.rb]: I added require in this place, to not force `active_job` presence for people that won't use this feature

File: lib/paperclip/has_attached_file.rb [hunk 3]
```diff
@@ -100,6 +101,20 @@ def add_active_record_callbacks
...
+      name = @name
+      if @klass.respond_to?(:after_commit)
```
Comment (morgoth) [inline lib/paperclip/has_attached_file.rb]: Should I care for case when there is no `after_commit` and add `after_save` then?

File: spec/paperclip/attachment_spec.rb [hunk 4]
```diff
@@ -1477,6 +1478,53 @@ def call(filename)
...
+  context "an attachment with process_in_background set" do
+    before do
+      ActiveJob::Base.queue_adapter = :test
+      ActiveJob::Base.logger = nil
+
+      rebuild_model styles: { thumb: "100x100" },
+                    only_process: [:none],
+                    process_in_background: [:thumb]
+      @file = File.new(fixture_file("5k.png"), "rb")
+      GlobalID.app = "paperclip-test"
```
Comment (morgoth) [inline spec/paperclip/attachment_spec.rb]: I couldn't make it work without it. In normal Rails app, this is set up by railties

File: spec/paperclip/attachment_spec.rb [hunk 5]
```diff
@@ -1477,6 +1478,53 @@ def call(filename)
     end
   end
 
+  context "an attachment with process_in_background set" do
+    before do
+      ActiveJob::Base.queue_adapter = :test
+      ActiveJob::Base.logger = nil
+
+      rebuild_model styles: { thumb: "100x100" },
+                    only_process: [:none],
```
Comment (morgoth) [inline spec/paperclip/attachment_spec.rb]: This is a hack to skip processing in real time. We need to think about better UI, so user can decide which styles he wants to be processed in real time/background

File: spec/paperclip/attachment_spec.rb [hunk 6]
```diff
@@ -1477,6 +1478,53 @@ def call(filename)
...
+      job = jobs[0]
+      assert_equal Paperclip::ProcessJob, job[:job]
+      assert_equal @dummy, GlobalID::Locator.locate(job[:args][0].values[0])
+      assert_equal "avatar", job[:args][1]
+      assert_equal "default", job[:queue]
+    end
+
+    it "does not process given styles in real time" do
+      attachment = @dummy.avatar
+      attachment.expects(:post_process).with(:none)
```
Comment (morgoth) [inline spec/paperclip/attachment_spec.rb]: I'm little confused when to use rspec and when minitest - which one is preferred?

**Comment by morgoth** [comment]
I'm also wondering how configuration for future features should look like, in example:
`has_attached_file :avatar, process_in_background: [:thumb], background_job_queue_name: "paperclip,  ...` or maybe group them all under one key?
`has_attached_file :avatar, process_in_background: {styles: [:thumb], queue_name: "paperclip"}}`

**Comment by andreiglingeanu** [comment]
@morgoth Let me know if I can help with something. Little bit of testing, anything. I'll be in touch.

**Comment by ScotterC** [comment]
As a maintainer of DelayedPaperclip, I'd love for this to be in Paperclip core. Much less hacky and easier to maintain in the future.

**Comment by morgoth** [comment]
@tute Do you think we can move it forward?
I think there are 2 blockers:
https://github.com/thoughtbot/paperclip/pull/2261#discussion_r71974153
https://github.com/thoughtbot/paperclip/pull/2261#issuecomment-234735329

**Comment by bdewater** [comment]
How would callbacks work if this is integrated? I have a potential use case that looks quite a lot like the one described here: https://github.com/jrgifford/delayed_paperclip/issues/176

File: spec/paperclip/attachment_spec.rb (hunk 6 above)
Comment (bdewater) [inline spec/paperclip/attachment_spec.rb]: RSpec, see message of 1f3a74678f459dc1b2d60efd49f3cfe4addc6f63

**Comment by Tonkpils** [comment]
any update on this PR?

**Comment by sidraval** [comment]
We are in the process of deprecating paperclip, and so are closing PRs that aren't immediate bug fixes or documentation updates. Thanks for your contribution, we regret not tending to it!

**Pull request closed without merging**